### PR TITLE
GH-34335: [C++][Parquet] Optimize Decoding DELTA_LENGTH_BYTE_ARRAY

### LINKED ISSUE #34335: [C++][Parquet] Improve the performance for Decoding DELTA_LENGTH_BYTE_ARRAY
### Describe the enhancement requested

To be honest, the previous logic is too slow. It's 20 time slower than PLAIN for ByteArray. I'll do some optimizations on it.

### Component(s)

C++, Parquet

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

According to https://github.com/apache/arrow/pull/34323 . DELTA_LENGTH_BYTE_ARRAY is much more slower. So do some optimizations.

### What changes are included in this PR?

Some tiny changes

### Are there any user-facing changes?

* Closes: #34335

### CODE REVIEW

**Comment by mapleFU** [comment]
@rok @wjones127 @wgtmac

File: cpp/src/parquet/encoding.cc [hunk 1]
```diff
@@ -2745,6 +2738,7 @@ class DeltaLengthByteArrayDecoder : public DecoderImpl,
     int32_t data_size = 0;
     const int32_t* length_ptr =
         reinterpret_cast<const int32_t*>(buffered_length_->data()) + length_idx_;
+    int bytes_left = decoder_->bytes_left();
```
Comment (wjones127) [inline cpp/src/parquet/encoding.cc:2741]: nit: could we rewrite this as `bytes_offset`? That way it's clearer why we are computing this before calling `decoder_->Advance()`

```suggestion
    int bytes_offset = len_ - decoder_->bytes_left();
```

File: cpp/src/parquet/encoding.cc [hunk 2]
```diff
@@ -2756,13 +2750,10 @@ class DeltaLengthByteArrayDecoder : public DecoderImpl,
...
     length_idx_ += max_values;
-
-    PARQUET_THROW_NOT_OK(buffered_data_->Resize(data_size));
-    if (decoder_->GetBatch(8, buffered_data_->mutable_data(), data_size) != data_size) {
+    if (ARROW_PREDICT_FALSE(!decoder_->Advance(8 * static_cast<int64_t>(data_size)))) {
       ParquetException::EofException();
     }
-    const uint8_t* data_ptr = buffered_data_->data();
-
+    const uint8_t* data_ptr = data_ + (len_ - bytes_left);
```
Comment (wjones127) [inline cpp/src/parquet/encoding.cc:2756]:
```suggestion
    const uint8_t* data_ptr = data_ + bytes_offset;
```

File: cpp/src/parquet/encoding.cc [hunk 3]
```diff
@@ -3071,8 +3061,14 @@ class DeltaByteArrayDecoder : public DecoderImpl,
     prefix_len_offset_ = 0;
     num_valid_values_ = num_prefix;
 
+    int bytes_left = decoder_->bytes_left();
+    int suffix_begins = len - bytes_left;
+    if (suffix_begins < 0) {
+      throw ParquetException(
+          "Wrong data in DeltaByteArrayDecoder: bytes_left more than length");
+    }
```
Comment (wjones127) [inline cpp/src/parquet/encoding.cc:3069]: If this happens, doesn't that mean we have already read past the buffer end (buffer overflow)? Seems like that should be handled in the lengths decoder.

**Review by wjones127 — CHANGES_REQUESTED** [review]

**Comment by mapleFU** [comment]
@wjones127 comment resolved, you can take a look when you have spare time :)

**Review by wjones127 — APPROVED** [review]

**Pull request merged**